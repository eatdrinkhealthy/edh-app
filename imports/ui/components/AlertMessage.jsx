// @flow
import Alert from "react-s-alert";

// note, css required for this package is imported in 'importPackagesCss.js'

class AlertMessage extends Alert {
  props: {
    position?: string,
    offset?: number,
    stack?: boolean | { limit: number },
  };

  // customizing default props (slightly different than Alert)
  static defaultProps = {
    position: "top-left",
    offset: 70,
    stack: { limit: 5 },
  };
}

export default AlertMessage;
