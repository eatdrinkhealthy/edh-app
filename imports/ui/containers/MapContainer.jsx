// @flow
import React, {
  Component,
} from "react";
import { connect } from "react-redux";
import { Meteor } from "meteor/meteor";
import _ from "lodash";
import AlertMessage from "../components/AlertMessage";
import Map from "../components/Map";
import { getLocation } from "../../utils/geoLocation";
import { setSearchResults } from "../../state/actions/searchResultsActions";
import { setSelectedVenue } from "../../state/actions/mapDisplayActions";

import type { IGoogleMapDisplay, ILatLng } from "google-map-react"; // eslint-disable-line import/first
import type { IVenue } from "../../state/reducers/searchResultsReducers";
import type { IState } from "../../state/stores/store";
import type { IEatDrinkFilter } from "../../state/reducers/eatDrinkFiltersReducers";
import type { IVenueTypeFilter } from "../../state/reducers/venueTypeFiltersReducers";

type IMapWrapperProps = {
  eatDrinkFilters: Array<IEatDrinkFilter>,
  venueTypeFilters: Array<IVenueTypeFilter>,
  searchResults: Array<IVenue>,
  setSearchResultsHandler: (searchResults: Array<IVenue>) => void,
  setSelectedVenueHandler: () => void,
  selectedVenueId: ?string,
};

type IMapWrapperState = {
  center: ILatLng,
  userLocation?: ILatLng,
  zoom: number,
};

export class MapWrapper extends Component {
  props: IMapWrapperProps;

  state: IMapWrapperState = {
    center: {
      lat: 32.789008,
      lng: -79.932115,
    },
    zoom: 3,
  };

  componentDidMount() {
    getLocation((position: Position) => {
      this.setState({
        center: {
          lat: position.coords.latitude,
          lng: position.coords.longitude,
        },
        userLocation: {
          lat: position.coords.latitude,
          lng: position.coords.longitude,
        },
        zoom: 15,
      });
    });
  }

  componentWillReceiveProps(nextProps: IMapWrapperProps) {
    if (this.filterHasChanged(nextProps)) {
      this.callFoursquareApi(nextProps.eatDrinkFilters, nextProps.venueTypeFilters);
    }
  }

  callFoursquareApi = (
    eatDrinkFilters: Array<IEatDrinkFilter>,
    venueTypeFilters: Array<IVenueTypeFilter>,
  ) => {
    const selectedEatDrinkFilters = eatDrinkFilters.filter(
      (filterItem: IEatDrinkFilter): boolean => (filterItem.on),
    );

    const selectedVenueTypeFilters = venueTypeFilters.filter(
      (filterItem: IVenueTypeFilter): boolean => (filterItem.on),
    );

    Meteor.call("getNearbyPlaces",
      {
        latitude: this.state.center.lat,
        longitude: this.state.center.lng,
        eatDrinkFilters: selectedEatDrinkFilters,
        venueTypeFilters: selectedVenueTypeFilters,
      },
      this.getNearbyPlacesCB,
    );
  };

  filterHasChanged = (nextProps: IMapWrapperProps): boolean => {
    const edfChanged = !_.isEqual(this.props.eatDrinkFilters, nextProps.eatDrinkFilters);
    const vtfChanged = !_.isEqual(this.props.venueTypeFilters, nextProps.venueTypeFilters);

    return edfChanged || vtfChanged;
  };

  // NOTE: this is an ES6 class property arrow function (preserves this context)
  getNearbyPlacesCB = (error: Error, result: Array<IVenue>) => {
    if (error) {
      AlertMessage.warning("Unable to search at this time...");
      // TODO potentially throw here (or confirm an exception is thrown by server)
    } else {
      if (!result.length) {
        // TODO consider only storing selected filters in state (instead of all
        //      filters) starting to repeat code to get which filters are selected
        const selectedEdFilters = this.props.eatDrinkFilters.filter(
          (filterItem: IEatDrinkFilter): boolean => (filterItem.on),
        );

        const selectedVtFilters = this.props.venueTypeFilters.filter(
          (filterItem: IVenueTypeFilter): boolean => (filterItem.on),
        );

        if (selectedEdFilters.length || selectedVtFilters.length) {
          AlertMessage.warning("No search results for current criteria...");
        }
      }
      // TODO potentially throw here (or log search criteria to a logger for evaluation)
      this.props.setSearchResultsHandler(result);
    }
  };

  handleMapChange = (mapChange: IGoogleMapDisplay) => {
    this.setState({
      center: {
        lat: mapChange.center.lat,
        lng: mapChange.center.lng,
      },
    });
  };

  render() { // eslint-disable-line flowtype/require-return-type
    return (
      <Map
        center={this.state.center}
        zoom={this.state.zoom}
        userLocation={this.state.userLocation}
        googleMapsApiKey={Meteor.settings.public.googleMapsApiKey}
        venues={this.props.searchResults}
        setSelectedVenueHandler={this.props.setSelectedVenueHandler}
        selectedVenueId={this.props.selectedVenueId}
        onMapChange={this.handleMapChange}
      />
    );
  }
}

type IStateProps = {
  eatDrinkFilters: Array<IEatDrinkFilter>,
  venueTypeFilters: Array<IVenueTypeFilter>,
  searchResults: Array<IVenue>,
  selectedVenueId: ?string,
};

const mapStateToProps = (state: IState): IStateProps => ({
  eatDrinkFilters: state.eatDrinkFilters,
  venueTypeFilters: state.venueTypeFilters,
  searchResults: state.searchResults,
  selectedVenueId: state.mapDisplay.selectedVenueId,
});

type IDispatchProps = {
  setSearchResultsHandler: (searchResults: Array<IVenue>) => void,
  setSelectedVenueHandler: (venueId: string) => void,
};

const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => ({
  setSearchResultsHandler: (searchResults: Array<IVenue>): void => (
    dispatch(setSearchResults(searchResults))
  ),
  setSelectedVenueHandler: (venueId: ?string): void => (
    dispatch(setSelectedVenue(venueId))
  ),
});

const MapContainer = connect(mapStateToProps, mapDispatchToProps)(MapWrapper);

export default MapContainer;
