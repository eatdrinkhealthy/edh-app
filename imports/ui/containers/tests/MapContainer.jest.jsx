// @flow
/* eslint-env jest */
/* eslint-disable func-names, prefer-arrow-callback, no-unused-expressions */
/* eslint-disable import/no-extraneous-dependencies, import/first */

// mock AlertMessage component (call to warning method)
// eslint-disable-next-line flowtype/require-return-type
jest.mock("../../components/AlertMessage", () => (
  class AlertMessage {
    static warning = jest.fn();
  }
));

import React from "react";
import { MemoryRouter } from "react-router-dom";
import { Provider } from "react-redux";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import _ from "lodash";
import { createStore } from "redux";
import appReducer from "../../../state/reducers";
import MapContainer, { MapWrapper } from "../MapContainer";
import sampleVenues from "../../../state/stores/tests/sampleVenueData";

/* eslint-disable no-duplicate-imports */
import type { IState } from "../../../state/stores/store";
import type { IEatDrinkFilter } from "../../../state/reducers/eatDrinkFiltersReducers";
/* eslint-enable no-duplicate-imports */

import AlertMessage from "../../components/AlertMessage";

describe("<MapWrapper />", function () {
  const edFilterList: Array<IEatDrinkFilter> = [
    { id: "jb1", name: "vegan", on: true, foursquareCategory: "1" },
    { id: "jb2", name: "raw", on: false, foursquareCategory: "2" },
    { id: "jb3", name: "juice", on: false, foursquareCategory: "3" },
  ];

  const noEdFilterList: Array<IEatDrinkFilter> = [
    { id: "jb1", name: "vegan", on: false, foursquareCategory: "1" },
    { id: "jb2", name: "raw", on: false, foursquareCategory: "2" },
    { id: "jb3", name: "juice", on: false, foursquareCategory: "3" },
  ];

  const vtFilterList: Array<IEatDrinkFilter> = [
    { id: "vt1", name: "Restaurant", on: true, foursquareCategory: "4" },
    { id: "vt2", name: "Coffee Shop", on: false, foursquareCategory: "5" },
    { id: "vt3", name: "Market", on: false, foursquareCategory: "6" },
  ];

  const noVtFilterList: Array<IEatDrinkFilter> = [
    { id: "vt1", name: "Restaurant", on: false, foursquareCategory: "4" },
    { id: "vt2", name: "Coffee Shop", on: false, foursquareCategory: "5" },
    { id: "vt3", name: "Market", on: false, foursquareCategory: "6" },
  ];

  const stubFn = jest.fn();

  const props = {
    eatDrinkFilters: edFilterList,
    venueTypeFilters: vtFilterList,
    searchResults: [],
    setSearchResultsHandler: stubFn,
    setSelectedVenueHandler: stubFn,
    selectedVenueId: null,
  };

  it("matches render snapshot", function () {
    // TODO - to capture more snapshot detail, use mount or react-test-renderer (BOTH FAIL HERE)
    const wrapper = shallow(<MapWrapper {...props} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should call AlertMessage.warning when calling getNearbyPlacesCB with an error", function () {
    const wrapper = shallow(<MapWrapper {...props} />);

    // $FlowFixMe (ignoring 'getNearbyPlacesCB' is not method of React$Component error)
    wrapper.instance().getNearbyPlacesCB("some error", undefined);
    expect(AlertMessage.warning).toHaveBeenCalledWith("Unable to search at this time...");
  });

  it("should call AlertMessage.warning when getNearbyPlacesCB has filters but no search results",
    function () {
      const wrapper = shallow(<MapWrapper {...props} />);

      // $FlowFixMe (ignoring 'getNearbyPlacesCB' is not method of React$Component error)
      wrapper.instance().getNearbyPlacesCB(undefined, []);
      expect(AlertMessage.warning).toHaveBeenCalledWith("No search results for current criteria...");
    });

  it("should NOT call AlertMessage.warning when getNearbyPlacesCB has no filters, no search results",
    function () {
      const propsNoFilters = {
        eatDrinkFilters: noEdFilterList,
        venueTypeFilters: noVtFilterList,
        searchResults: [],
        setSearchResultsHandler: stubFn,
        setSelectedVenueHandler: stubFn,
        selectedVenueId: null,
      };

      const wrapper = shallow(<MapWrapper {...propsNoFilters} />);

      // $FlowFixMe (ignoring 'getNearbyPlacesCB' is not method of React$Component error)
      wrapper.instance().getNearbyPlacesCB(undefined, []);
      expect(AlertMessage.warning).not.toHaveBeenCalled();
    });

  it("should know when filter has NOT changed", function () {
    const nextProps = _.cloneDeep(props);    // make a copy of props

    const wrapper = shallow(<MapWrapper {...props} />);
    // $FlowFixMe (ignoring 'filterHasChanged' is not method of React$Component error)
    expect(wrapper.instance().filterHasChanged(nextProps)).toBe(false);
  });

  it("should know when filter has changed", function () {
    const nextProps = _.cloneDeep(props);        // make a copy of props
    nextProps.venueTypeFilters[0].on = false;    // change something in it

    const wrapper = shallow(<MapWrapper {...props} />);
    // $FlowFixMe (ignoring 'filterHasChanged' is not method of React$Component error)
    expect(wrapper.instance().filterHasChanged(nextProps)).toBe(true);
  });
});

describe("<MapContainer />", function () {
  const testDefaultState: IState = {
    eatDrinkFilters: [
      { id: "vegan", name: "Vegan", on: true, foursquareCategory: "4" },
      { id: "raw", name: "Raw", on: false, foursquareCategory: "5" },
      { id: "juice", name: "Juice", on: false, foursquareCategory: "6" },
    ],
    venueTypeFilters: [
      { id: "coffeeShop", name: "Coffee Shop", on: true, foursquareCategory: "7" },
      { id: "grocery", name: "Market / Store", on: false, foursquareCategory: "8" },
      { id: "healthFoodStore", name: "Health Food Store", on: false, foursquareCategory: "9" },
    ],
    searchResults: [
      sampleVenues[3],
      sampleVenues[4],
      sampleVenues[5],
    ],
    mapDisplay: { selectedVenueId: "B" },
  };

  const testStore = createStore(
    appReducer,
    testDefaultState,
  );

  const wrapper = mount(
    <Provider store={testStore}>
      <MemoryRouter>
        <MapContainer />
      </MemoryRouter>
    </Provider>,
  );

  it("should set eatDrinkFilters for MapWrapper from state", function () {
    expect(wrapper.find("MapWrapper").at(0).props().eatDrinkFilters)
      .toEqual(testDefaultState.eatDrinkFilters);
  });

  it("should set venueTypeFilters for MapWrapper from state", function () {
    expect(wrapper.find("MapWrapper").at(0).props().venueTypeFilters)
      .toEqual(testDefaultState.venueTypeFilters);
  });

  it("should set search results for MapWrapper from state", function () {
    expect(wrapper.find("MapWrapper").at(0).props().searchResults)
      .toEqual(testDefaultState.searchResults);
  });

  it("should set selectedVenueId from state", function () {
    expect(wrapper.find("MapWrapper").at(0).props().selectedVenueId)
      .toEqual(testDefaultState.mapDisplay.selectedVenueId);
  });
});
