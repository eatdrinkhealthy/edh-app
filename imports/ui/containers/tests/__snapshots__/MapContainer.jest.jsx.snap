exports[`<MapComponent /> matches render snapshot 1`] = `
<div>
  <Navbar />
  <LocationsMap
    center={
      Object {
        "lat": 32.789008,
        "lng": -79.932115,
      }
    }
    googleMapsApiKey="TestAPIKeyValue"
    venues={Array []}
    zoom={16} />
</div>
`;
